hugegraph-api/src/main/java/com/baidu/hugegraph/api/auth/ProjectAPI.java

```diff
@@ -0,0 +1,235 @@
+import com.baidu.hugegraph.auth.AuthManager;
+import com.baidu.hugegraph.auth.HugeProject;
+import com.baidu.hugegraph.backend.id.Id;
+import com.baidu.hugegraph.core.GraphManager;
+import com.baidu.hugegraph.define.Checkable;
+import com.baidu.hugegraph.exception.NotFoundException;
+import com.baidu.hugegraph.server.RestServer;
+import com.baidu.hugegraph.util.E;
+import com.baidu.hugegraph.util.Log;
+import com.codahale.metrics.annotation.Timed;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Strings;
+
+@Path("graphs/{graph}/auth/projects")
+@Singleton
+public class ProjectAPI extends API {
+
+    private static final Logger LOG = Log.logger(RestServer.class);
+    private static final String ACTION_ADD_GRAPH = "add_graph";
+    private static final String ACTION_REMOVE_GRAPH = "remove_graph";
+
+    @POST
+    @Timed
+    @StatusFilter.Status(StatusFilter.Status.CREATED)
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    public String create(@Context GraphManager manager,
+                         @PathParam("graph") String graph,
+                         JsonProject jsonProject) {
+        LOG.debug("Graph [{}] create project: {}", graph, jsonProject);
+        checkCreatingBody(jsonProject);
+
+        HugeGraph g = graph(manager, graph);
+        HugeProject project = jsonProject.build();
+        Id projectId = manager.authManager().createProject(project);
+        /*
+         * Some fields of project(like admin_group) can only be known after
+         * created
+         */
+        project = manager.authManager().getProject(projectId);
+        return manager.serializer(g).writeAuthElement(project);
+    }
+
+    @PUT
+    @Timed
+    @Path("{id}")
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    public String update(@Context GraphManager manager,
+                         @PathParam("graph") String graph,
+                         @PathParam("id") String id,
+                         @QueryParam("action") String action,
+                         JsonProject jsonProject) {
+        LOG.debug("Graph [{}] update {} project: {}", graph, action,
+                  jsonProject);
+
+        HugeGraph g = graph(manager, graph);
+        HugeProject project;
+        Id projectId = UserAPI.parseId(id);
+        AuthManager authManager = manager.authManager();
+        try {
+            project = authManager.getProject(projectId);
+        } catch (NotFoundException e) {
+            throw new IllegalArgumentException("Invalid project id: " + id);
+        }
+        project = jsonProject.build(action, project);
+        if (this.isAddGraph(action)) {
+            authManager.updateProjectAddGraph(projectId, jsonProject.graph);
+        } else if (this.isRemoveGraph(action)) {
+            authManager.updateProjectRemoveGraph(projectId, jsonProject.graph);
+        } else {
+            authManager.updateProject(project);
```

Comment:
checkUpdate

Reply:
The check logic is in the method "JsonProject.build", so there is no need to call the method "checkUpdate" here